test/e2e/gateway_api_test.go

```diff
@@ -423,6 +428,286 @@ func testGatewayAPIRBAC(t *testing.T) {
 	}
 }
 
+func testGatewayAPIDNS(t *testing.T) {
+	domain := "gws." + dnsConfig.Spec.BaseDomain
+
+	gatewayClass, err := createGatewayClass(operatorcontroller.OpenShiftDefaultGatewayClassName, operatorcontroller.OpenShiftGatewayClassControllerName)
+	if err != nil {
+		t.Fatalf("failed to create gatewayclass: %v", err)
+	}
+
+	testCases := []struct {
+		name           string
+		createGateways []struct {
+			gatewayName string
+			namespace   string
+			hostname    *string
+		}
+		expectedListenerConditions []metav1.Condition
+		expectedDNSRecords         map[expectedDnsRecord]bool
+	}{
+		// TODO: In this case Gateway Listeners should be reported as conflicted. To be fixed in the future release.
+		{
+			name: "multipleGatewaysSameListenerHostname",
+			createGateways: []struct {
+				gatewayName string
+				namespace   string
+				hostname    *string
+			}{
+				{gatewayName: "gw1", namespace: operatorcontroller.DefaultOperandNamespace, hostname: ptr.To("abc." + domain)},
+				{gatewayName: "gw2", namespace: operatorcontroller.DefaultOperandNamespace, hostname: ptr.To("abc." + domain)},
+			},
+			expectedListenerConditions: []metav1.Condition{
+				{Type: "Accepted", Status: metav1.ConditionTrue},
+				{Type: "Conflicted", Status: metav1.ConditionFalse},
+				{Type: "Programmed", Status: metav1.ConditionTrue},
+				{Type: "ResolvedRefs", Status: metav1.ConditionTrue},
+			},
+			expectedDNSRecords: map[expectedDnsRecord]bool{
+				{dnsName: "abc." + domain + ".", gatewayName: "gw1"}: true,
+				{dnsName: "abc." + domain + ".", gatewayName: "gw2"}: true,
+			},
+		},
+		{
+			name: "gatewayListenersWithOverlappingHostname",
+			createGateways: []struct {
+				gatewayName string
+				namespace   string
+				hostname    *string
+			}{
+				{gatewayName: "gw3", namespace: operatorcontroller.DefaultOperandNamespace, hostname: ptr.To("qwe." + domain)},
+				{gatewayName: "gw4", namespace: operatorcontroller.DefaultOperandNamespace, hostname: ptr.To("*." + domain)},
+			},
+			expectedListenerConditions: []metav1.Condition{
+				{Type: "Accepted", Status: metav1.ConditionTrue},
+				{Type: "Conflicted", Status: metav1.ConditionFalse},
+				{Type: "Programmed", Status: metav1.ConditionTrue},
+				{Type: "ResolvedRefs", Status: metav1.ConditionTrue},
+			},
+			expectedDNSRecords: map[expectedDnsRecord]bool{
+				{dnsName: "qwe." + domain + ".", gatewayName: "gw3"}: true,
+				{dnsName: "*." + domain + ".", gatewayName: "gw4"}:   true,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var gateways []*gatewayapiv1.Gateway
+
+			// Create gateways
+			for _, gateway := range tc.createGateways {
+				createdGateway, err := createGatewayWithListeners(gatewayClass, gateway.gatewayName, gateway.namespace, []struct {
+					Name     string
+					Hostname *string
+				}{
+					{Name: "http", Hostname: gateway.hostname},
+				})
```

Comment:
nit: The anonymous struct parameter in the function call feels awkward to me, but there's probably nothing technically wrong with it.  Ordinarily you could do something like this first and use the gw parameter:
`gw := []struct{{Name:"http", Hostname: gateway.hostname},}
`